proposals/0197-chilli-peppers.md

```diff
@@ -0,0 +1,147 @@
+A new transaction level resource requesting the transfer of state from cold to hot
+Block level constraints on the total cold to hot state transition
+These added features will allow for a pricing market for the bandwidth from cold state to hot state (via priority fees), and allow block producers to more optimally pack blocks to get the highest possible throughput on already hot state and constrain access to cold state to be within bounds that the validator network will be able to keep up with.
+ 
+ 
+ 
+## Motivation
+ 
+In commodity hardware (for fixed cost), there is a fundamental tradeoff between the the size of accessible state and the bandwidth of random access to that state. On-chip caches >> RAM >> SSD >> HDD >> NAS increase by orders of magnitude in size, while falling by orders of magnitude in bandwidth.
+For Solana (or any blockchain), treating all state as equivalent (regardless of its usage patterns) means that either total state size will
+be limited by the size of RAM, or the throughput of the network will be limited to the bandwidth of disks. Actual usage patterns
+(and expectations for future usage patterns as the network grows) show that a relatively small amount of the total state is accessed
+frequently, and most of the state is accessed infrequently.
+This usage pattern allows a hot/cold tiered state design to allow the total state size available from disk, while achieving the
+throughput available from RAM.
+ 
+## New Terminology
+ 
+Chili Peppers (State Units?) - a consumable resource representing the number of bytes loaded into the "hot store" (active, frequently
+accessed memory) for state operations on the Solana blockchain. Each Chili Pepper corresponds directly to one byte of data loaded, providing
+a precise mechanism to quantify and limit the resources consumed by transactions in terms of state memory usage.
+Note that Chili Peppers required for a transaction are not the size of hot state touched by the transaction, but rather the amount of cold
+state made hot or new state allocated. Transactions that are only accessing already hot state would require 0 Chili Peppers.
+ 
+Block Chili Pepper Limit - The maximum number of Chili Peppers that can be requested in a single block.
+ 
+Block Chili Pepper Clock - a cumulative measure of the total Chili Peppers requested since the genesis of the blockchain. At the beginning of each block, the Block State Unit Clock is updated to reflect the cumulative total of Chili Peppers consumed up to that point, ensuring a monotonically increasing record of state consumption.
+ 
+Account Chili Pepper Clock Timestamp - each account in the "hot" state needs to keep track of the value of the Block Chili Pepper Clock the last time it was accessed (read or written to). This allows determination of which accounts are hot and which are cold.
+ 
+Hot Cache Size - A new predefined constant, termed "Hot Cache Size" corresponds approximately to the size of the hot state supported
```

Comment:
How is the size determined?

Reply:
Most likely by deciding what the "standard" size of a solana network box should be and backing into how much should be used for account state.   For example, given the small amount of new-state actually referenced per block, I would guess we could set this at 50gb and it would be 99% good...

Reply:
Thinking a bit more. "hot cache size" (50gb) is the physical memory capacity, right?

chili pepper meter measures just the bytes of the account's data loaded from disk. But there are other extra cost in addition to the account's data bytes to load, such as account's index, account's chili-pepper store etc. The actual physical memory will be larger than the chili pepper bytes?

If we are targeting 50gb physical memory, should we take into accounts of those extra cost and reduce them from the physical memory budget? If so, how much should we reduce?